// bio.io/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.version = exports.xhr = exports.seqs = exports.parser = exports.newick = exports.matrix = exports.gff = exports.fasta = exports.clustal = undefined;

var _clustal2 = require("./clustal");

var _clustal3 = _interopRequireDefault(_clustal2);

var _fasta2 = require("./fasta");

var _fasta3 = _interopRequireDefault(_fasta2);

var _gff2 = require("./gff");

var _gff3 = _interopRequireDefault(_gff2);

var _matrix2 = require("./matrix");

var _matrix3 = _interopRequireDefault(_matrix2);

var _newick2 = require("./newick");

var _newick3 = _interopRequireDefault(_newick2);

var _parser2 = require("./parser");

var _parser3 = _interopRequireDefault(_parser2);

var _seqs2 = require("./seqs");

var _seqs3 = _interopRequireDefault(_seqs2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.clustal = _clustal3.default; // TODO: make BLAST slim
//export blast from "./blast";

exports.fasta = _fasta3.default;
exports.gff = _gff3.default;
exports.matrix = _matrix3.default;
exports.newick = _newick3.default;
exports.parser = _parser3.default;
exports.seqs = _seqs3.default;

// convenience export

var xhr = require("xhr");
exports.xhr = xhr;

// version will be automatically injected by webpack
// IO_VERSION is only defined if loaded via webpack

var VERSION = "imported";
if (typeof IO_VERSION !== "undefined") {
    VERSION = IO_VERSION;
}

var version = exports.version = VERSION;
